crates/mem-pool/src/sync/p2p.rs

```diff
@@ -0,0 +1,465 @@
+                        );
+                        shared.subscribers.insert(context.id);
+                        tracing::info!(
+                            subscribers.len = shared.subscribers.len(),
+                            added = context.id.value(),
+                        );
+                        break;
+                    } else {
+                        let try_again_block =
+                            shared.buffer.first_block_buffered().unwrap_or_default();
+                        drop(shared); // Unlock as soon as possible.
+                        let try_again = TryAgain::new_builder()
+                            .block_number(try_again_block.1.pack())
+                            .block_hash(try_again_block.0.pack())
+                            .build();
+                        let reply = P2PSyncResponse::new_builder()
+                            .set(P2PSyncResponseUnion::TryAgain(try_again))
+                            .build();
+                        warn_result(
+                            control
+                                .send_message_to(
+                                    context.id,
+                                    P2P_MEM_BLOCK_SYNC_PROTOCOL,
+                                    reply.as_bytes(),
+                                )
+                                .await,
+                        );
+                    }
+                }
+                while let Some(Ok(_)) = read_part.next().await {}
+                let mut shared = shared.lock().await;
+                shared.subscribers.remove(&context.id);
+                tracing::info!(
+                    subscribers.len = shared.subscribers.len(),
+                    removed = context.id.value(),
+                );
+            });
+        }))
+        .build()
+}
+
+pub(crate) fn sync_server_publisher(
+    control: ServiceAsyncControl,
+    shared: Arc<Mutex<SyncServerState>>,
+) -> Publisher {
+    Publisher { control, shared }
+}
+
+async fn wait_for_tip(
+    mem_pool: &Mutex<MemPool>,
+    tip: (H256, u64),
+    at_least_once: bool,
+) -> Result<(), (H256, u64)> {
+    let mem_pool = mem_pool.lock().await;
+    let current = mem_pool.current_tip();
+    if current == tip {
+        return Ok(());
+    }
+    if current.1 > tip.1 && !at_least_once {
+        return Err(current);
+    }
+    let mut receiver = mem_pool.subscribe_new_tip();
+    drop(mem_pool);
+    tracing::info!("waiting for tip update");
+    loop {
+        match receiver.recv().await {
+            Ok(new_tip) => {
+                tracing::info!(new_tip = %HashAndNumber::from(new_tip));
+                if new_tip == tip {
+                    return Ok(());
+                }
+                if new_tip.1 > tip.1 {
+                    return Err(new_tip);
+                }
+            }
+            Err(broadcast::error::RecvError::Lagged(_)) => {}
+            Err(_) => panic!("tip update sender closed"),
```

Comment:
This line is not reachable, because we are holding an reference to the mem_pool, which holds the sender, so the sender won't be closed. 

But we do need to consider how does this affect graceful shutdown.

Reply:
I have added shutdown support for sync client protocol spawned tasks and tested that they work as expected.

(Sync server does not need any special handling because it only blocks on stream read so it ends naturally when we shutdown the whole p2p service and because it does not hold a reference to the mem pool (or any other important resources) so even if it does not shutdown it does not matter.)